Assets/Scripts/RobotProgramming/EngineLogic/BaseEngineLogic.cs

```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using UnityEngine;
+
+namespace Cosmobot.Api
+{
+    [DisallowMultipleComponent]
+    public class BaseEngineLogic : MonoBehaviour, IEngineLogic
+    {
+        private ManualResetEvent taskCompletedEvent;
+        private CancellationToken cancellationToken;
+        private ProgrammableFunctionWrapper wrapper;
+
+        public void SetupThread(ManualResetEvent taskEvent, CancellationToken token, ConcurrentQueue<Action> commandQueue)
+        {
+            taskCompletedEvent = taskEvent;
+            cancellationToken = token;
+            wrapper = new ProgrammableFunctionWrapper(taskCompletedEvent, cancellationToken, commandQueue);
+        }
+
+        public IReadOnlyDictionary<string, Delegate> GetFunctions()
+        {
+            //Expose robot's ingame functions here
+            return new Dictionary<string, Delegate>()
+            {
+                { "wait", wrapper.Wrap<float>(Wait)},
+                { "log", wrapper.Wrap<string>(Log)},
+                { "logWarning", wrapper.Wrap<string>(LogWarning)},
+                { "logError", wrapper.Wrap<string>(LogError)},
+                { "dance", wrapper.Wrap(Dance)},
+            };
+        }
+
+        //FUNCTION IMPLEMENTATIONS
+        // Implement Robot's functions here
+        // functions must only return void, primitives or types in Cosmobot.Api.Types
+        // functions also must have a unique name
+        // (!)remember to expose functions ingame in Dictionary above
+        // (!)remember to call "taskCompletedEvent.Set();" when yours code is finished or robot will wait infinitely
+        private void Wait(float seconds)
+        {
+            StartCoroutine(WaitCoroutine(seconds));
+        }
+
+        private IEnumerator WaitCoroutine(float seconds)
+        {
+            yield return new WaitForSeconds(seconds);
+            taskCompletedEvent.Set();
+        }
+
+        private void Log(string message)
+        {
+            Debug.Log(message);
+            taskCompletedEvent.Set();
+        }
+
+        public void LogInternal(string message)
+        {
+            Debug.Log(message);
+        }
```

Comment:
nitpick: keep "Internal" methods together and because they are public i would suggest to keep them before "robot functions" or maybe on the top - before "SetupThread"